src/mpid/ch4/netmod/ofi/ofi_am_impl.h

```diff
@@ -473,4 +531,126 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_inject(int rank,
     goto fn_exit;
 }
 
+MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_do_am_isend_pipeline(int rank, MPIR_Comm * comm,
+                                                            int handler_id, const void *am_hdr,
+                                                            size_t am_hdr_sz, const void *buf,
+                                                            size_t count, MPI_Datatype datatype,
+                                                            MPIR_Request * sreq, size_t data_sz)
+{
+    int dt_contig, mpi_errno = MPI_SUCCESS;
+    char *send_buf;
+    MPI_Aint packed_size;
+    MPI_Aint dt_true_lb, send_size;
+    bool need_packing = false;
+    MPIDI_OFI_am_send_pipeline_request_t *send_req;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_DO_AM_ISEND_PIPELINE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_DO_AM_ISEND_PIPELINE);
+
+    MPIDI_Datatype_check_contig_lb(datatype, count, dt_contig, dt_true_lb);
+    send_buf = (char *) buf + dt_true_lb;
+
+    send_size = MPIDI_OFI_DEFAULT_SHORT_SEND_SIZE - am_hdr_sz - sizeof(MPIDI_OFI_am_header_t);
+    send_size = MPL_MIN(send_size, data_sz);
+    MPIDIG_send_init(sreq, datatype, data_sz);
+    MPIDIG_send_start(sreq);
+
+    need_packing = dt_contig ? false : true;
+
+    MPL_pointer_attr_t attr;
+    MPIR_GPU_query_pointer_attr(buf, &attr);
+    if (attr.type == MPL_GPU_POINTER_DEV && !MPIDI_OFI_ENABLE_HMEM) {
+        /* Force packing of GPU buffer in host memory */
+        need_packing = true;
+    }
+
+
+    if (MPIDI_OFI_global.deferred_am_isend_q) {
+        /* if the deferred queue is not empty, all new ops must be deferred to maintain ordering */
+        goto fn_deferred;
+    }
```

Comment:
Similar to my comment to the eager send, can we add a flag to avoid some code duplication with `issue_deferred_am_isend_pipeline`?

Reply:
I don't think so. The codes of eager and pipeline differ at multiple places. Plus, they were one function with a parameter controlling eager v.s. pipeline. You wanted them to be separated.

Reply:
see above

Reply:
It already very early. You mean before the set on of send_size?

Reply:
They are merged.